src/main/java/com/google/devtools/build/lib/actions/cache/MetadataInjector.java

```diff
@@ -0,0 +1,49 @@
+// Copyright 2018 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package com.google.devtools.build.lib.actions.cache;
+
+import com.google.devtools.build.lib.actions.Artifact;
+import com.google.devtools.build.lib.actions.Artifact.SpecialArtifact;
+import com.google.devtools.build.lib.actions.FileArtifactValue.RemoteFileArtifactValue;
+import com.google.devtools.build.lib.vfs.PathFragment;
+import java.util.Map;
+
+/**
+ * Allows to inject metadata of action outputs into skyframe.
+ */
+public interface MetadataInjector {
```

Comment:
A simpler approach would be to change SpawnExecutionContext.getMetadataProvider to getMetadataHandler. I actually looked at some internal code today that I'd like to remove, but it would require the SpawnRunner to have access to the MetadataHandler...

Reply:
As discussed offline, MetadaHandler contains *a lot* of methods that should be internal to SkyFrame. I believe MetadataInjector together with MetadataProvider is what should be exposed outside of SkyFrame. As discussed offline, you need the `markOmitted` method from MetadataHandler and I believe this method is a good candidate to be moved to MetadataInjector as marking an artifact as omitted is also a way of injecting metadata about it. WDYT?